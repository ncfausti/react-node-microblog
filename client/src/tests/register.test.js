/* eslint-disable no-undef */
import React from 'react';
import {
  fireEvent, render, waitFor, screen,
} from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import 'isomorphic-fetch';
import fetchMock from 'jest-fetch-mock';
import Register from '../components/register';

fetchMock.enableMocks();

describe('webpage correctly handles submit & shows msg from server', () => {
  test('display error msg when server rejects this registration', async () => {
    fetch.mockResponseOnce(JSON.stringify({
      status: 'err',
      msg: '✖ Registration failed: Invalid username or password provided.',
    }));

    const { getByRole } = render(<Register.WrappedComponent/>);
    fireEvent.click(getByRole('button', { name: 'Register' }));

    await waitFor(() => {
      expect(screen.getByText('✖ Registration failed: Invalid username or password provided.')).toBeInTheDocument();
    });
  });

  test('mock redirect is called when registration was successful', async () => {
    const mockPush = jest.fn().mockImplementation(() => {});
    const mockRouterProps = { push: mockPush };
    fetch.mockResponseOnce(JSON.stringify({
      status: 'ok',
      msg: 'Registration was successfull.',
    }));

    const { getByRole } = render(<Register.WrappedComponent history={mockRouterProps}/>);
    fireEvent.click(getByRole('button', { name: 'Register' }));

    await waitFor(() => {
      expect(mockPush).toHaveBeenCalledWith({
        pathname: '/login',
        state: {
          msg: '✔ Registration was successful. Please login using your crendentials.',
        },
      });
    });
  });

  test('Page matches snapshot', () => {
    const component = renderer.create(
      <Register.WrappedComponent />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
