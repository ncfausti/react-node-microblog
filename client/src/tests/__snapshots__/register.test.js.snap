// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webpage correctly handles submit & shows msg from server Page matches snapshot 1`] = `
<div
  className="root"
>
  <div
    className="topbar"
  >
    <a
      href="/"
      id="title"
    >
      MicroBlog
    </a>
    <div
      className="topbar-right"
    >
      <a
        href="/login"
        id="login"
      >
        Login
      </a>
      <a
        href="/register"
        id="signup-active"
      >
        Signup
      </a>
    </div>
  </div>
  <div
    id="errMsg"
  >
    <span>
      
    </span>
  </div>
  <div
    className="register"
  >
    <form
      onSubmit={[Function]}
    >
      <label>
        Username
      </label>
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
      <label>
        Passsword
      </label>
      <input
        onChange={[Function]}
        type="password"
        value=""
      />
      <label>
        Email
      </label>
      <input
        onChange={[Function]}
        placeholder="Optional"
        type="text"
        value=""
      />
      <label>
        Nickname
      </label>
      <input
        onChange={[Function]}
        placeholder="Optional"
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Register
      </button>
    </form>
  </div>
</div>
`;
